astar_path_planner/src/astar_path_planner.cpp

```diff
@@ -0,0 +1,186 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#include "astar_path_planner/astar_path_planner.hpp"
+#include <memory>
+#include <set>
+#include <vector>
+#include <queue>
+
+namespace astar_path_planner
+{
+
+void AStarPathPlanner::DeclareParameters(rclcpp_lifecycle::LifecycleNode::SharedPtr node)
+{
+  node->declare_parameter("goal_threshold", 0.05);
+  node->declare_parameter("grid_size", 0.01);
+}
+
+AStarPathPlanner::AStarPathPlanner(
+  rclcpp_lifecycle::LifecycleNode::SharedPtr node,
+  std::shared_ptr<nav2_costmap_2d::Costmap2DROS> ros_costmap)
+: logger_(node->get_logger()),
+  ros_costmap_(ros_costmap)
+{
+  goal_threshold_ = node->get_parameter("goal_threshold").as_double();
+  grid_size_ = node->get_parameter("grid_size").as_double();
+}
+
+std::vector<Point> AStarPathPlanner::Plan(const Point & start, const Point & goal)
+{
+  std::set<Point, PointComparator> expanded{PointComparator{}};
+  std::priority_queue<FrontierEntry, std::vector<FrontierEntry>,
+    FrontierEntryComparator> frontier{FrontierEntryComparator{}};
+
+  goal_ = goal;
+
+  std::vector<Point> initial_path = {start};
+  frontier.push({initial_path, GetPathCost(initial_path)});
+
+  while (!frontier.empty()) {
+    const auto [path, cost] = frontier.top();
+    frontier.pop();
+    const auto last_state = path.back();
+
+    if (expanded.find(last_state) == expanded.end()) {
+      expanded.insert(last_state);
+
+      if (IsGoal(last_state)) {
+        std::vector<Point> reduced_path{path};
+        ReducePath(reduced_path);
+        return reduced_path;
+      }
+
+      const auto neighbors = GetAdjacentPoints(last_state);
+
+      for (const auto & neighbor : neighbors) {
+        std::vector<Point> new_path(path);
+        new_path.push_back(neighbor);
+        const auto cost = GetPathCost(new_path);
+        frontier.push({new_path, cost});
+      }
+    }
+  }
+
+  throw std::runtime_error("Could not find path!");
+}
+
+bool AStarPathPlanner::IsGoal(const Point & point)
+{
+  return (point - goal_).norm() < goal_threshold_;
+}
+
+std::vector<Point> AStarPathPlanner::GetAdjacentPoints(const Point & point)
+{
+  std::vector<Point> neighbors;
+
+  for (auto dx = -grid_size_; dx <= grid_size_; dx += grid_size_) {
+    for (auto dy = -grid_size_; dy <= grid_size_; dy += grid_size_) {
+      if (std::abs(dx) < 1e-4 && std::abs(dy) < 1e-4) {
+        continue;
+      }
+      const Point neighbor{point.x() + dx, point.y() + dy};
+      if (!PointInCollision(neighbor)) {
+        neighbors.push_back(neighbor);
+      }
+    }
+  }
+  return neighbors;
+}
+
+double AStarPathPlanner::GetPathCost(const std::vector<Point> & path)
```

Comment:
This doesn't have to be calculated from scratch each time. You can use the previous cost once you remove the heuristic and only do a single step.

Reply:
True. I can switch it. My only hesitation is whether that's trickier to explain to the students.

Reply:
I'll file this under showing them what they should be doing rather than what is immediately easy to understand. I wouldn't want to see this in team code so I think we should not show it.